### PR TITLE
Laravel tutorial, fixes #388

### LINKED ISSUE #388: Laravel Starter Config
Split from https://github.com/TugboatQA/docs/issues/137:

Create pages for the following "starter kits" from the TugboatQA repo:

Laravel: [TugboatQA/Laravel-Starter-Kit](https://github.com/TugboatQA/Laravel-Starter-Kit)

### PR DESCRIPTION
**Pull request by penyaskito**
Laravel tutorial, fixes #388

### CODE REVIEW

File: content/starter-configs/tutorials/laravel.md [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+        - find "${TUGBOAT_ROOT}/storage" -type d -exec chmod 2775 {} \;
+        - find "${TUGBOAT_ROOT}/storage" -type f -exec chmod 0664 {} \;
+        # Compile vite templates
+        - npm install
+        - npm run build
+
+  database:
+    # Use the latest available version of MariaDB by not specifying a
+    # version
+    image: tugboatqa/mariadb:10.5
```
Comment (q0rban) [inline content/starter-configs/tutorials/laravel.md:133]: Should there be a database to import, or is that happening in the update phase of the webserver service?

File: content/starter-configs/tutorials/laravel.md [hunk 2]
```diff
@@ -0,0 +1,153 @@
...
+        - php artisan config:cache
+        # Run any pending migrations.
+        - php artisan migrate --force
+        # Ensure storage permissions
+        - chgrp -R www-data "${TUGBOAT_ROOT}/storage"
+        - find "${TUGBOAT_ROOT}/storage" -type d -exec chmod 2775 {} \;
+        - find "${TUGBOAT_ROOT}/storage" -type f -exec chmod 0664 {} \;
+        # Compile vite templates
+        - npm install
+        - npm run build
```
Comment (q0rban) [inline content/starter-configs/tutorials/laravel.md:128]: Since the `update` build phase only runs on base previews or previews with no base, I think you will want a `build` phase as well with some of these commands. For example, something like this? (would artisan migrate run on every deployment typically?)

```suggestion
        - npm run build
      build:
        - composer install --optimize-autoloader
        # Clear caches.
        - php artisan config:cache
        # Run any pending migrations.
        - php artisan migrate --force
        - npm install
        - npm run build
```

**Review by q0rban — COMMENTED** [review]
@penyaskito this is _fantastic_, thank you so much for working on this! 👏👏👏

**Comment by q0rban** [comment]
FYI, I pushed changes from `yarn lint-fix` to this PR.

File: content/starter-configs/tutorials/laravel.md (hunk 1 above)
Comment (penyaskito) [reply content/starter-configs/tutorials/laravel.md:134]: Laravel provides ways to easily create test data.

That's done on the init:
```
        # Generate the db structure with some data.
        - php artisan migrate --seed
```

`artisan migrate` generates the structure, and `--seed` is a flag for invoking seeders.

In your project you will have some classes tagged as seeders that generate that test data.

File: content/starter-configs/tutorials/laravel.md (hunk 2 above)
Comment (penyaskito) [reply content/starter-configs/tutorials/laravel.md:129]: Thanks! Yes, migrations should run on every deployment.

File: content/starter-configs/tutorials/laravel.md (hunk 1 above)
Comment (penyaskito) [reply content/starter-configs/tutorials/laravel.md:134]: I'm adding more verbose comments clarifying this just in case.

File: content/starter-configs/tutorials/laravel.md [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        # Run any pending migrations. This will ensure your data has the last
+        # migrations applied.
+        - php artisan migrate --force
+        # Ensure storage permissions
+        - chgrp -R www-data "${TUGBOAT_ROOT}/storage"
+        - find "${TUGBOAT_ROOT}/storage" -type d -exec chmod 2775 {} \;
+        - find "${TUGBOAT_ROOT}/storage" -type f -exec chmod 0664 {} \;
+        # Compile vite templates
+        - npm install
+        - npm run build
```
Comment (q0rban) [inline content/starter-configs/tutorials/laravel.md:135]:
```suggestion
        - chgrp -R www-data "${TUGBOAT_ROOT}/storage"
        - find "${TUGBOAT_ROOT}/storage" -type d -exec chmod 2775 {} \;
        - find "${TUGBOAT_ROOT}/storage" -type f -exec chmod 0664 {} \;
```

I don't think these are needed in both update and build, but correct me if I'm wrong, I have no laravel experience. ☺️

Comment (penyaskito) [reply content/starter-configs/tutorials/laravel.md:135]: tbh I've read https://docs.tugboatqa.com/building-a-preview/preview-deep-dive/how-previews-work/ 30 times and still not 100% sure what goes where.

I'm pretty clear what goes on init, but build and update are quite fuzzy for me.
If I'm right, build would run in every case (having or not a base preview) 
Update will happen if I manually refresh a preview.  But it still uses the base preview if any (or init step if none).

Basically, what we need to do is run migrate any time where there might be a code change, as a new migration (script with sequence of changes to a previous db state) could be added. In laravel those are build incrementally (e.g. add this new table, or add this new column to existing table).

Comment (penyaskito) [reply content/starter-configs/tutorials/laravel.md:135]: And to be clear, there is no harm on running migrate twice, as it detects which migrations are already applied.

Comment (q0rban) [reply content/starter-configs/tutorials/laravel.md:143]: You're not alone, it is confusing. When a preview is built with no base preview, it runs each of the command groups in order, `init`, `update`, and then `build`. Then the state of the services is committed back to a docker image. If a new preview is built from a Base preview, it then creates containers from those docker images with the saved state, and just runs the `build` phase. 

Based on what you've said here, I think my suggested change is correct. We only want to run `migrate` once per preview on every build.

Comment (penyaskito) [reply content/starter-configs/tutorials/laravel.md:148]: Then I wonder if our `update` should be just empty (`build` will run in every case), or some stuff from `init` should be on `update`. But I'm pretty sure that we want the background tasks installed on init, but that will trigger them and they require the db to be already initialized, which requires e.g. a previous composer install.

Comment (q0rban) [reply content/starter-configs/tutorials/laravel.md]: okay, so reading through the commands a bit more carefully, I think maybe you should move the `artisan migrate --seed` into the `update` phase. The update phase is what will be run on a Base Preview whenever a *Refresh* occurs.


```
    commands:
      # Commands that set up the basic preview infrastructure
      init:
        # Install opcache and mod-rewrite.
        - docker-php-ext-install opcache
        - a2enmod headers rewrite

        # Install node
        - apt-get update
        - apt-get install -yq ca-certificates curl gnupg
        - mkdir -p /etc/apt/keyrings
        - curl -fsSL https://deb.nodesource.com/gpgkey/nodesource-repo.gpg.key | gpg --dearmor -o
          /etc/apt/keyrings/nodesource.gpg
        - echo "deb [signed-by=/etc/apt/keyrings/nodesource.gpg] https://deb.nodesource.com/node_18.x nodistro main" |
          tee /etc/apt/sources.list.d/nodesource.list
        - apt-get update
        - apt-get install -yq nodejs

        # Copy the tugboat .env we created before
        - cp "${TUGBOAT_ROOT}/.tugboat/.env.tugboat" "${TUGBOAT_ROOT}/.env"

        # Link the document root to the expected path. This example links
        # /public to the docroot
        - ln -snf "${TUGBOAT_ROOT}/public" "${DOCROOT}"

        # Ensure storage permissions
        - chgrp -R www-data "${TUGBOAT_ROOT}/storage"
        - find "${TUGBOAT_ROOT}/storage" -type d -exec chmod 2775 {} \;
        - find "${TUGBOAT_ROOT}/storage" -type f -exec chmod 0664 {} \;

        # Install the workers
        - mkdir -p /etc/service/webserver
        - cp .tugboat/etc/service/webserver/run /etc/service/webserver/run
        - chmod +x /etc/service/webserver/run
        
      update:
        # Composer install
        - composer install --optimize-autoloader
        # Generate the random key if you didn't set APP_KEY before.
        - php artisan key:generate
        # Generate the db structure with some data. Here you have several choices:
        # 1. If you have seeders on your project, you might just run migrate with --seed.
        # - php artisan migrate --seed
        # 2. You might want just the structure without any test data.
        # - php artisan migrate
        # 3. Or you might load a database dump from somewhere else. That's up to you.
        - php artisan migrate --seed

      build:
        - composer install --optimize-autoloader
        # Clear caches.
        - php artisan config:cache
        # Run any pending migrations. This will ensure your data has the last
        # migrations applied.
        - php artisan migrate --force
        # Compile vite templates
        - npm install
        - npm run build
```

Is there an easy demo laravel app we could create? I would love to have a demo repo for this at https://github.com/TugboatDemo. Then we can hone the above commands with a "real" project.

**Comment by q0rban** [comment]
I wonder if @mglaman would have some suggestions here as well, as he has experience with both laravel and tugboat

File: content/starter-configs/tutorials/laravel.md (hunk 3 above)
Comment (q0rban) [reply content/starter-configs/tutorials/laravel.md]: > I agree with you, but if I understood etc/service/webserver/run will be running before update happens, and that would fail because it requires the db to exist already.

It will fail, but it should still allow the build to complete, at which point it will stop all containers, and then commit them to images, and recreate the containers and start them.

Comment (penyaskito) [reply content/starter-configs/tutorials/laravel.md]: Not that you needed it, but I confirm it worked.

**Review by q0rban — APPROVED** [review]
